### PR TITLE
Discretionary glyphs vs. hyphenation

### LINKED ISSUE #355: Drop characters when hyphenating
So the feature added at my request in #265 to setup custom hyphenation patterns is great for words that are exceptions to the normal rules, but I'm finding it leaves me hanging when it comes to the actual rules.

I've recently discovered that there is more than one style guide for how to hyphenate words with internal apostrophes in Turkish. My personal preference is to place the hyphenation points after the apostrophe, but this is not what all publishing houses want to see. Many of them require the apostrophe to be _dropped_ in the event that the word gets hyphenated at that point. For example

> müjdesinin

might normally hyphenate as

```
müj-de-si-nin
```

However in the event it is used as a proper noun the suffixes (after the third person possessive) are set apart with an apostrophe:

> Müjdesi’nin

```
Müj-de-si’-nin
```

However in the event that a hyphenation point is actually used, the apostrophe should go away.

> Mesih'in Müjdesi-
> nin tek gerçek

The problem here is two fold.
1. There are a huge number of words where this could be an issue, so adding case by case rules is cumbersome. My current solution for getting hyphens after apostrophes rather than before is to search the full text of my books for _all_ the apostrophized words and manually add them to an exception list with the pre-apostrophe hyphen points removed.
2. The current exception handler does not provide a mechanism for the input and output strings to be different. The input of the function is the hyphenated form and the function assumes the word is the text string with hyphens removed. This won't work to setup a publishing workflow that conforms to the expectation of apostrophes being dropped if used as hyphenation points.

### PR DESCRIPTION
**Pull request by alerque**
Assuming I can sort out concessionaires as hyphenation content...

Fixes #355
Does not touch #583 since that uses manual discretionaries and this uses hyphenation patterns.
Exacerbated by and might further exacerbate #1362, but will be more correct and make that bug easier to squash.

~~Not ready for merge yet (since it currently just blindly assumes all languages need this treatment) but some preliminary hacking on hyphenation for Turkish that handles apostrophes as discretionaries if they fall at hyphenation points (which they are by definition).~~

### CODE REVIEW

File: core/nodefactory.lua [hunk 1]
```diff
@@ -276,6 +275,11 @@ function nodefactory.discretionary:toText ()
 end
 
 function nodefactory.discretionary:outputYourself (typesetter, line)
+  -- TODO this is an indication of a deeper bug. If we were asked to output
+  -- discretionaries but the parent nnode is not hyphenated then we're
+  -- outputting things that were only used for measuring "what if" scenarios in
+  -- break point calculations. These nodes shouldn't exist at this point.
+  if self.parent and not self.parent.hyphenated then return end
```
Comment (Omikhleia) [inline core/nodefactory.lua:282]: Re: "These nodes shouldn't exist at this point": There's nothing (currently) to remove unused discretionaries. The last one in a line is marked as used (the next test below), in `computeLineRatio` which also sets the parent's hyphenated to true in that case.

Comment (alerque) [reply core/nodefactory.lua:282]: I wasn't expected them to be *removed* necesesarily, but they are getting added at some points they should not be. For example somehow after typesetting a line we're ending up with hyphenation discretionaries *after* each word. A word with 3 hyphenation points that isn't even a candidate for hyphenation will end up with 3 discretionaries *after* the word (not in their places). The actual words near the end of a line that are hyphenation candidates based on having filled up a line and gone over-length will have the correct placement.

**Comment by alerque** [comment]
This gets the job done, but one thing I'm not happy with is the blunt-force special casing for Turkish. AT the moment Turkish is the only language with hyphenation patterns with apostrophes but I have to imagine it shouldn't be. At the very least some other Turkic languages are likely to need similar treatment. I have to imagine there are other languages with...

* intraword apostrophes or other marks
* that can have segments longer than 2 letters on the right hand side
* and the marks need special treatment if landing on a (used) hyphenation point.

That being said we don't have a good ergonomic place to hang such a setting on yet. I suppose eventually a per-language setting and a couple of hooks are in order.

I think I'm going to move on without those things at this point because I'm not in a good position to come up with something ergonomic and just hacking one in before refactoring the setting scope (#1367 and friends) and making the typesetter, line breaker, and language support bits proper modules is going to materially slow down the line break loops for all languages for no gain.

File: core/nodefactory.lua (hunk 1 above)
Comment (alerque) [reply core/nodefactory.lua:282]: I split out my findings on this issue in #1528.

**Pull request merged**